src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java

```diff
@@ -0,0 +1,218 @@
+        super(configuration);
+    }
+
+    /**
+     * This function will validate if the supplied atlas object is valid for the check.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return {@code true} if this object should be checked
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge && Edge.isMasterEdgeIdentifier(object.getIdentifier())
+                && !HighwayTag.isLinkHighway(object) && VALID_HIGHWAY_TAG.test(object)
+                && !JunctionTag.isRoundabout(object);
+    }
+
+    /**
+     * This is the actual function that will check to see whether the object needs to be flagged.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+        final EdgePredicate edgePredicate = new EdgePredicate(edge);
+        final Set<Edge> inEdges = edge.inEdges().stream().filter(this::validCheckForObject)
+                .filter(edgePredicate::isSameHeading).collect(Collectors.toSet());
+        final Set<Edge> outEdges = edge.outEdges().stream().filter(this::validCheckForObject)
+                .filter(edgePredicate::isSameHeading).collect(Collectors.toSet());
+        if (inEdges.isEmpty() || outEdges.isEmpty())
+        {
+            return Optional.empty();
+        }
+
+        final Set<String> matchingInAndOutEdgeNames = getMatchingInAndOutEdgeNames(inEdges,
+                outEdges);
+
+        if (matchingInAndOutEdgeNames.isEmpty())
+        {
+            // There is no pair of inedge and out edge with same name.
+            return Optional.empty();
+        }
+
+        // Create flag when we have in edge and out edge with same name but intermediate edge
+        // doesn't have a name.
+        if (!edge.getName().isPresent())
+        {
+            return Optional.of(
+                    createFlag(object, this.getLocalizedInstruction(0, edge.getOsmIdentifier())));
+        }
+        final Optional<String> edgeName = edge.getName();
+
+        if (edgeName.isPresent() && !matchingInAndOutEdgeNames.contains(edgeName.get()))
+        {
+            // Case 1: Edge name: Tai. Incoming edge names: Tai , Shai. Outgoing edge name: Shai.
+            // Case 2: Edge name: Tai. Incoming edge names: Shai. Outgoing edge name: Tai, Shai.
+            // Case 3: Edge name: Tai. Incoming edges names: Tai, Shai. Outgoing edge names: Shai,
+            // Pendler.
+            final Set<Edge> connectedEdges = edge.connectedEdges().stream()
+                    .filter(this::validCheckForObject).collect(Collectors.toSet());
+            if (findMatchingEdgeNameWithConnectedEdges(connectedEdges, edgeName.get()))
```

Comment:
Here it looks like we don't flag edges if one of its in/out edges shares its name. If that's the case I think this ```if``` block can be moved to line 113 or so since it doesn't seem necessary to wait for getMatchingInAndOutEdgeNames() to finish.

Reply:
I want to do this validation only when there is at least one matching in and out edge name. Because, there was a failing case where there is matching in and out edge but the interconnecting edge is connected to another edge which has a different name and which doesn't have either incoming or outgoing edge for it.